libs/utils/src/tabulation/contest_filtering.ts

```diff
@@ -0,0 +1,260 @@
+    const ballotStyleContestIds = ballotStyleContestIdsLookup[ballotStyle.id];
+    assert(ballotStyleContestIds);
+    for (const associatedPrecinctId of ballotStyle.precincts) {
+      const associatedPrecinctContestIds = lookup[associatedPrecinctId];
+      assert(associatedPrecinctContestIds);
+      for (const contestId of ballotStyleContestIds) {
+        associatedPrecinctContestIds.add(contestId);
+      }
+    }
+  }
+  return lookup;
+}
+
+export const getContestIdsForPrecinct = createElectionMetadataLookupFunction(
+  buildPrecinctContestIdsLookup
+);
+
+export function getBallotStyleIdsForFilter(
+  electionDefinition: ElectionDefinition,
+  filter?: Tabulation.Filter
+): Set<BallotStyleId> {
+  const { election } = electionDefinition;
+
+  let ballotStyleIds = new Set(election.ballotStyles.map((bs) => bs.id));
+  if (!filter) return ballotStyleIds;
+
+  // Ballot Style, Party, and Precinct filters can all narrow down contests
+
+  // narrow down by explicit Ballot Style filter
+  if (filter.ballotStyleIds) {
+    ballotStyleIds = intersectSets([
+      ballotStyleIds,
+      new Set(filter.ballotStyleIds),
+    ]);
+  }
+
+  // narrow down by Party filter
+  if (filter.partyIds) {
+    const { partyIds } = filter;
+    const ballotStyleIdsRestrictedByParty: Set<BallotStyleId> = unionSets(
+      partyIds.map((partyId) => {
+        const ballotStyles = getBallotStylesByPartyId(
+          electionDefinition,
+          partyId
+        );
+        return new Set(ballotStyles.map((bs) => bs.id));
+      })
+    );
+
+    ballotStyleIds = intersectSets([
+      ballotStyleIds,
+      ballotStyleIdsRestrictedByParty,
+    ]);
+  }
+
+  // narrow down by Precinct filter
+  if (filter.precinctIds) {
+    const { precinctIds } = filter;
+    const ballotStyleIdsRestrictedByPrecinct: Set<BallotStyleId> = unionSets(
+      precinctIds.map((precinctId) => {
+        const ballotStyles = getBallotStylesByPrecinctId(
+          electionDefinition,
+          precinctId
+        );
+        return new Set(ballotStyles.map((bs) => bs.id));
+      })
+    );
+
+    ballotStyleIds = intersectSets([
+      ballotStyleIds,
+      ballotStyleIdsRestrictedByPrecinct,
+    ]);
+  }
+
+  return ballotStyleIds;
+}
+
+function getContestIdsForBallotStyleIds(
+  electionDefinition: ElectionDefinition,
+  ballotStyleIds: BallotStyleId[]
+): Set<ContestId> {
+  return unionSets(
+    ballotStyleIds.map((ballotStyleId) =>
+      getContestIdsForBallotStyle(electionDefinition, ballotStyleId)
+    )
+  );
+}
+
+/**
+ * Filters can restrict the ballot styles that can appear in a report - if there
+ * are no ballot styles in a report, it's not a valid report.
+ */
+export function getContestIdsForFilter(
+  electionDefinition: ElectionDefinition,
+  filter?: Tabulation.Filter
+): Set<ContestId> {
+  return getContestIdsForBallotStyleIds(electionDefinition, [
+    ...getBallotStyleIdsForFilter(electionDefinition, filter),
+  ]);
+}
```

Comment:
The idea here is that a consumer, handling a report that is both filtered and split, could `getContestIdsForFilter` once and then `getContestIdsForSplit` for each split and intersect the two sets to get the actual contest IDs.